deepsubtree.go

```diff
@@ -0,0 +1,271 @@
+package iavl
+
+import (
+	"bytes"
+	"fmt"
+)
+
+// Represents a IAVL Deep Subtree that can contain
+// a subset of nodes of an IAVL tree
+type DeepSubTree struct {
+	*MutableTree
+}
+
+// Adds the node with the given key in the Deep Subtree
+// using the given full IAVL tree along everything along
+// the path of that node
+func (dst *DeepSubTree) AddPath(tree *ImmutableTree, key []byte) error {
+	path, val, err := tree.root.PathToLeaf(tree, key)
+	if err != nil {
+		return err
+	}
+
+	err = dst.addPath(path, val)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Helper method to add given leaf node in the Deep Subtree
+// using the given PathToLeaf
+func (dst *DeepSubTree) addPath(pl PathToLeaf, leaf *Node) error {
+	hash, err := leaf._hash()
+	if err != nil {
+		return err
+	}
+
+	n := NewNode(leaf.key, leaf.value, leaf.version)
+	n._hash()
+	err = dst.ndb.SaveNode(n)
+	if err != nil {
+		return err
+	}
+
+	prevHash := n.hash
+	for i := len(pl) - 1; i >= 0; i-- {
+		pin := pl[i]
+		hash, err = pin.Hash(hash)
+		if err != nil {
+			return err
+		}
+		if pin.Left == nil {
+			pin.Left = prevHash
+		} else if pin.Right == nil {
+			pin.Right = prevHash
+		}
+		n := &Node{
+			subtreeHeight: pin.Height,
+			size:          pin.Size,
+			version:       pin.Version,
+			leftHash:      pin.Left,
+			rightHash:     pin.Right,
+			hash:          hash,
+		}
+		prevHash = hash
+
+		has, err := dst.ndb.Has(n.hash)
+		if err != nil {
+			return err
+		}
+		if !has {
+			err = dst.ndb.SaveNode(n)
+			if err != nil {
+				return err
+			}
+		}
+		if i == 0 {
+			if err := dst.ndb.Commit(); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// Traverses in the nodes in the NodeDB in the Deep Subtree
+// and links them together using the populated left and right
+// hashes and sets the root to be the node with the given rootHash
+func (dst *DeepSubTree) BuildTree(rootHash []byte) error {
+	nodes, traverseErr := dst.ndb.nodes()
+	if traverseErr != nil {
+		return fmt.Errorf("could not traverse nodedb: %w", traverseErr)
+	}
+	// Traverse through nodes and link them correctly
+	for _, node := range nodes {
+		pnode, _ := dst.ndb.GetNode(node.hash)
+		if len(pnode.leftHash) > 0 && pnode.leftNode == nil {
+			pnode.leftNode, _ = dst.ndb.GetNode(pnode.leftHash)
+		}
+		if len(pnode.rightHash) > 0 && pnode.rightNode == nil {
+			pnode.rightNode, _ = dst.ndb.GetNode(pnode.rightHash)
+		}
+	}
+	// Now that nodes are linked correctly, traverse again
+	// and set their keys correctly
+	for _, node := range nodes {
+		pnode, _ := dst.ndb.GetNode(node.hash)
+		if pnode.leftNode != nil {
+			pnode.key = pnode.leftNode.getHighestKey()
+		}
+
+		if pnode.rightNode != nil {
+			pnode.key = pnode.rightNode.getLowestKey()
+		}
+	}
+	if dst.root == nil {
+		rootNode, rootErr := dst.ndb.GetNode(rootHash)
+		if rootErr != nil {
+			return fmt.Errorf("could not set root of deep subtree: %w", rootErr)
+		}
+		dst.root = rootNode
+	}
+
+	return nil
```

Comment:
Is there a reason this isn't first? Seems like a quit early case if we can't get the root node. 

Is there a reason we have to do all the node traversing before this step?

Also, should we verify that in the case that `dst.root != nil` that the root node hash matches the provided hash?

Reply:
Great catch, thank you! Updated.